packages/edit-navigation/src/store/actions.js

```diff
@@ -0,0 +1,274 @@
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import {
+	getNavigationPost,
+	getPendingActions,
+	isProcessingPost,
+	getMenuItemToClientIdMapping,
+	resolveMenuItems,
+	dispatch,
+	apiFetch,
+} from './controls';
+import { menuItemsQuery } from './utils';
+
+/**
+ * Creates a menu item for every block that doesn't have an associated menuItem.
+ * Requests POST /wp/v2/menu-items once for every menu item created.
+ *
+ * @param {Object} post A navigation post to process
+ * @return {Function} An action creator
+ */
+export const createMissingMenuItems = serializeProcessing( function* ( post ) {
+	const menuId = post.meta.menuId;
+
+	const mapping = yield {
+		type: 'GET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
+		postId: post.id,
+	};
+	const clientIdToMenuId = invert( mapping );
+
+	const stack = [ post.blocks[ 0 ] ];
+	while ( stack.length ) {
+		const block = stack.pop();
+		if ( ! ( block.clientId in clientIdToMenuId ) ) {
+			const menuItem = yield apiFetch( {
+				path: `/__experimental/menu-items`,
+				method: 'POST',
+				data: {
+					title: 'Placeholder',
+					url: 'Placeholder',
+					menu_order: 0,
+				},
+			} );
+
+			mapping[ menuItem.id ] = block.clientId;
+			const menuItems = yield resolveMenuItems( menuId );
+			yield dispatch(
+				'core',
+				'receiveEntityRecords',
+				'root',
+				'menuItem',
+				[ ...menuItems, menuItem ],
+				menuItemsQuery( menuId ),
+				false
+			);
+		}
+		stack.push( ...block.innerBlocks );
+	}
+
+	yield {
+		type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
+		postId: post.id,
+		mapping,
+	};
+} );
+
+/**
+ * Converts all the blocks into menu items and submits a batch request to save everything at once.
+ *
+ * @param {Object} post A navigation post to process
+ * @return {Function} An action creator
+ */
+export const saveNavigationPost = serializeProcessing( function* ( post ) {
+	const menuId = post.meta.menuId;
+	const menuItemsByClientId = mapMenuItemsByClientId(
+		yield resolveMenuItems( menuId ),
+		yield getMenuItemToClientIdMapping( post.id )
+	);
+
+	try {
+		const response = yield* batchSave(
+			menuId,
+			menuItemsByClientId,
+			post.blocks[ 0 ]
+		);
+		if ( ! response.success ) {
+			throw new Error();
+		}
+		yield dispatch(
+			'core/notices',
+			'createSuccessNotice',
+			__( 'Navigation saved.' ),
+			{
+				type: 'snackbar',
+			}
+		);
+	} catch ( e ) {
+		yield dispatch(
+			'core/notices',
+			'createErrorNotice',
+			__( 'There was an error.' ),
+			{
+				type: 'snackbar',
+			}
+		);
+	}
+} );
+
+function mapMenuItemsByClientId( menuItems, clientIdsByMenuId ) {
+	const result = {};
+	if ( ! menuItems || ! clientIdsByMenuId ) {
+		return result;
+	}
+	for ( const menuItem of menuItems ) {
+		const clientId = clientIdsByMenuId[ menuItem.id ];
+		if ( clientId ) {
+			result[ clientId ] = menuItem;
+		}
+	}
+	return result;
+}
+
+function* batchSave( menuId, menuItemsByClientId, navigationBlock ) {
+	const { nonce, stylesheet } = yield apiFetch( {
```

Comment:
Need to handle any error that this request may have.

Reply:
There are no scenarios when the endpoint returns `success === false` so the only type of error that could happen here is one that would cause `apiFetch` to throw, in which case it would be handled by the caller. I added a short sanity check to throw if the returned nonce is falsy. Later on we should distinguish between different types of errors and maybe even point specific menu items that triggered them.